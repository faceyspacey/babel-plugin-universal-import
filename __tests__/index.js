/* eslint-disable no-template-curly-in-string */

const pluginTester = require('babel-plugin-tester')
const createBabylonOptions = require('babylon-options')
const plugin = require('../index')
const babel = require('@babel/core')

const babelOptions = {
  filename: '/dev/null',
  parserOpts: createBabylonOptions({
    plugins: ['dynamicImport']
  }),
  presets: ['@babel/preset-react'],
  plugins: [
    '@babel/plugin-syntax-dynamic-import',
    ['@babel/plugin-transform-modules-commonjs', { strictMode: false }]
  ],
  babelrc: false
}

pluginTester({
  plugin,
  babelOptions,
  snapshot: true,
  tests: {
    'static import': 'import("./Foo")',
    'static import (with relative paths)': 'import("../../Foo")',
    'static import (with file extension)': 'import("./Foo.js")',
    'static import (string template)': 'import(`./base`)',
    'static import (string template + relative paths)': 'import(`../../base`)',
    'static import (import as function with relative paths + nested folder)':
      'const obj = {component:()=>import(`../components/nestedComponent`)}; ()=> obj.component()',
    'static import (relative paths + nested folder)':
      'import(`../components/nestedComponent`)',
    'dynamic import (string template)': 'import(`./base/${page}`)',
    'dynamic import (string template with nested folder)':
      'import(`./base/${page}/nested/folder`)',
    'dynamic import (string template with multiple nested folders)':
      'import(`./base/${page}/nested/{$another}folder`)',
    'dynamic import (string template - dynamic at 1st segment)':
      'import(`./${page}`)',
    'dynamic import (string template + relative paths)':
      'import(`../../base/${page}`)',
    'await import() should receive a thennable without calling .then':
      'async ({ page }) => await import(`../components/${page}`);',
    'babelServer: true': {
      code: 'import("./Foo")',
      pluginOptions: { babelServer: true }
    },
    'disableWarnings: true': {
      code: 'import("./Foo")',
      pluginOptions: { disableWarnings: true }
    },
    'existing chunkName': 'import(/* webpackChunkName: \'Bar\' */"./Foo")',
    'multiple imports': 'import("one"); import("two"); import("three");'
  }
})

// toggle from test.skip to test.only when working on the plugin using Wallaby
test.skip('wallaby-live-coding', () => {
  // const input = 'async ({ page }) => await import(`../components/${page}`);'
  // const input = 'import("../../Foo.js")'
  // const input = 'universal(props => import(`./footer/${props.experiment}`));'
  const input = 'import(`./base/${page}/index`)'

  const output = babel.transform(
    input,
    Object.assign({}, babelOptions, {
      plugins: babelOptions.plugins.concat([plugin])
    })
  )

  expect(output.code).toBeDefined()
})
