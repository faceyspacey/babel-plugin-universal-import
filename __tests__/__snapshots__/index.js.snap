// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`await import() should receive a thennable without calling .then 1`] = `
"
async ({ page }) => await import(\`../components/\${page}\`);

      ↓ ↓ ↓ ↓ ↓ ↓

var _universalImport = _interopRequireDefault(require(\\"babel-plugin-universal-import/universalImport\\")).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

async ({
  page
}) => await _universalImport(() => import(
/* webpackChunkName: 'components/[request]' */
\`../components/\${page}\`), false);
"
`;

exports[`babelServer: true 1`] = `
"
import(\\"./Foo\\")

      ↓ ↓ ↓ ↓ ↓ ↓

var _path = _interopRequireDefault(require(\\"path\\")).default;

var _universalImport = _interopRequireDefault(require(\\"babel-plugin-universal-import/universalImport\\")).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

_universalImport({
  id: \\"./Foo\\",
  file: \\"currentFile.js\\",
  path: () => _path.join(__dirname, \\"./Foo\\"),
  resolve: () => require.resolveWeak(\\"./Foo\\"),
  chunkName: () => \\"Foo\\"
});
"
`;

exports[`disableWarnings: true 1`] = `
"
import(\\"./Foo\\")

      ↓ ↓ ↓ ↓ ↓ ↓

var _path = _interopRequireDefault(require(\\"path\\")).default;

var _universalImport = _interopRequireDefault(require(\\"babel-plugin-universal-import/universalImport\\")).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

_universalImport({
  id: \\"./Foo\\",
  file: \\"currentFile.js\\",
  load: () => import(
  /* webpackChunkName: 'Foo' */
  \\"./Foo\\"),
  path: () => _path.join(__dirname, \\"./Foo\\"),
  resolve: () => require.resolveWeak(\\"./Foo\\"),
  chunkName: () => \\"Foo\\"
});
"
`;

exports[`dynamic import (string template + relative paths) 1`] = `
"
import(\`../../base/\${page}\`)

      ↓ ↓ ↓ ↓ ↓ ↓

var _path = _interopRequireDefault(require(\\"path\\")).default;

var _universalImport = _interopRequireDefault(require(\\"babel-plugin-universal-import/universalImport\\")).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

_universalImport({
  id: \\"../../base/\${page}\\",
  file: \\"currentFile.js\\",
  load: () => import(
  /* webpackChunkName: 'base/[request]' */
  \`../../base/\${page}\`),
  path: () => _path.join(__dirname, \`../../base/\${page}\`),
  resolve: () => require.resolveWeak(\`../../base/\${page}\`),
  chunkName: () => \`base/\${page}\`
});
"
`;

exports[`dynamic import (string template - dynamic at 1st segment) 1`] = `
"
import(\`./\${page}\`)

      ↓ ↓ ↓ ↓ ↓ ↓

var _path = _interopRequireDefault(require(\\"path\\")).default;

var _universalImport = _interopRequireDefault(require(\\"babel-plugin-universal-import/universalImport\\")).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

_universalImport({
  id: \\"./\${page}\\",
  file: \\"currentFile.js\\",
  load: () => import(
  /* webpackChunkName: '[request]' */
  \`./\${page}\`),
  path: () => _path.join(__dirname, \`./\${page}\`),
  resolve: () => require.resolveWeak(\`./\${page}\`),
  chunkName: () => \`\${page}\`
});
"
`;

exports[`dynamic import (string template with multiple nested folders) 1`] = `
"
import(\`./base/\${page}/nested/{$another}folder\`)

      ↓ ↓ ↓ ↓ ↓ ↓

var _path = _interopRequireDefault(require(\\"path\\")).default;

var _universalImport = _interopRequireDefault(require(\\"babel-plugin-universal-import/universalImport\\")).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

_universalImport({
  id: \\"./base/\${page}/nested/{$another}folder\\",
  file: \\"currentFile.js\\",
  load: () => import(
  /* webpackChunkName: 'base/[request]' */
  \`./base/\${page}/nested/{$another}folder\`),
  path: () => _path.join(__dirname, \`./base/\${page}/nested/{$another}folder\`),
  resolve: () => require.resolveWeak(\`./base/\${page}/nested/{$another}folder\`),
  chunkName: () => \`base/\${page}-nested-{$another}folder\`
});
"
`;

exports[`dynamic import (string template with nested folder) 1`] = `
"
import(\`./base/\${page}/nested/folder\`)

      ↓ ↓ ↓ ↓ ↓ ↓

var _path = _interopRequireDefault(require(\\"path\\")).default;

var _universalImport = _interopRequireDefault(require(\\"babel-plugin-universal-import/universalImport\\")).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

_universalImport({
  id: \\"./base/\${page}/nested/folder\\",
  file: \\"currentFile.js\\",
  load: () => import(
  /* webpackChunkName: 'base/[request]' */
  \`./base/\${page}/nested/folder\`),
  path: () => _path.join(__dirname, \`./base/\${page}/nested/folder\`),
  resolve: () => require.resolveWeak(\`./base/\${page}/nested/folder\`),
  chunkName: () => \`base/\${page}-nested-folder\`
});
"
`;

exports[`dynamic import (string template) 1`] = `
"
import(\`./base/\${page}\`)

      ↓ ↓ ↓ ↓ ↓ ↓

var _path = _interopRequireDefault(require(\\"path\\")).default;

var _universalImport = _interopRequireDefault(require(\\"babel-plugin-universal-import/universalImport\\")).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

_universalImport({
  id: \\"./base/\${page}\\",
  file: \\"currentFile.js\\",
  load: () => import(
  /* webpackChunkName: 'base/[request]' */
  \`./base/\${page}\`),
  path: () => _path.join(__dirname, \`./base/\${page}\`),
  resolve: () => require.resolveWeak(\`./base/\${page}\`),
  chunkName: () => \`base/\${page}\`
});
"
`;

exports[`existing chunkName 1`] = `
"
import(/* webpackChunkName: 'Bar' */\\"./Foo\\")

      ↓ ↓ ↓ ↓ ↓ ↓

var _path = _interopRequireDefault(require(\\"path\\")).default;

var _universalImport = _interopRequireDefault(require(\\"babel-plugin-universal-import/universalImport\\")).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

_universalImport({
  id: \\"./Foo\\",
  file: \\"currentFile.js\\",
  load: () => import(
  /* webpackChunkName: 'Bar' */
  \\"./Foo\\"),
  path: () => _path.join(__dirname, \\"./Foo\\"),
  resolve: () => require.resolveWeak(\\"./Foo\\"),
  chunkName: () => \\"Bar\\"
});
"
`;

exports[`static import (import as function with relative paths + nested folder) 1`] = `
"
const obj = {component:()=>import(\`../components/nestedComponent\`)}; ()=> obj.component()

      ↓ ↓ ↓ ↓ ↓ ↓

var _path = _interopRequireDefault(require(\\"path\\")).default;

var _universalImport = _interopRequireDefault(require(\\"babel-plugin-universal-import/universalImport\\")).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const obj = {
  component: () => _universalImport({
    id: \\"../components/nestedComponent\\",
    file: \\"currentFile.js\\",
    load: () => import(
    /* webpackChunkName: 'components-nestedComponent' */
    \`../components/nestedComponent\`),
    path: () => _path.join(__dirname, \`../components/nestedComponent\`),
    resolve: () => require.resolveWeak(\`../components/nestedComponent\`),
    chunkName: () => \\"components-nestedComponent\\"
  })
};

() => obj.component();
"
`;

exports[`static import (relative paths + nested folder) 1`] = `
"
import(\`../components/nestedComponent\`)

      ↓ ↓ ↓ ↓ ↓ ↓

var _path = _interopRequireDefault(require(\\"path\\")).default;

var _universalImport = _interopRequireDefault(require(\\"babel-plugin-universal-import/universalImport\\")).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

_universalImport({
  id: \\"../components/nestedComponent\\",
  file: \\"currentFile.js\\",
  load: () => import(
  /* webpackChunkName: 'components-nestedComponent' */
  \`../components/nestedComponent\`),
  path: () => _path.join(__dirname, \`../components/nestedComponent\`),
  resolve: () => require.resolveWeak(\`../components/nestedComponent\`),
  chunkName: () => \\"components-nestedComponent\\"
});
"
`;

exports[`static import (string template + relative paths) 1`] = `
"
import(\`../../base\`)

      ↓ ↓ ↓ ↓ ↓ ↓

var _path = _interopRequireDefault(require(\\"path\\")).default;

var _universalImport = _interopRequireDefault(require(\\"babel-plugin-universal-import/universalImport\\")).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

_universalImport({
  id: \\"../../base\\",
  file: \\"currentFile.js\\",
  load: () => import(
  /* webpackChunkName: 'base' */
  \`../../base\`),
  path: () => _path.join(__dirname, \`../../base\`),
  resolve: () => require.resolveWeak(\`../../base\`),
  chunkName: () => \`base\`
});
"
`;

exports[`static import (string template) 1`] = `
"
import(\`./base\`)

      ↓ ↓ ↓ ↓ ↓ ↓

var _path = _interopRequireDefault(require(\\"path\\")).default;

var _universalImport = _interopRequireDefault(require(\\"babel-plugin-universal-import/universalImport\\")).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

_universalImport({
  id: \\"./base\\",
  file: \\"currentFile.js\\",
  load: () => import(
  /* webpackChunkName: 'base' */
  \`./base\`),
  path: () => _path.join(__dirname, \`./base\`),
  resolve: () => require.resolveWeak(\`./base\`),
  chunkName: () => \`base\`
});
"
`;

exports[`static import (with file extension) 1`] = `
"
import(\\"./Foo.js\\")

      ↓ ↓ ↓ ↓ ↓ ↓

var _path = _interopRequireDefault(require(\\"path\\")).default;

var _universalImport = _interopRequireDefault(require(\\"babel-plugin-universal-import/universalImport\\")).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

_universalImport({
  id: \\"./Foo.js\\",
  file: \\"currentFile.js\\",
  load: () => import(
  /* webpackChunkName: 'Foo' */
  \\"./Foo.js\\"),
  path: () => _path.join(__dirname, \\"./Foo.js\\"),
  resolve: () => require.resolveWeak(\\"./Foo.js\\"),
  chunkName: () => \\"Foo\\"
});
"
`;

exports[`static import (with relative paths) 1`] = `
"
import(\\"../../Foo\\")

      ↓ ↓ ↓ ↓ ↓ ↓

var _path = _interopRequireDefault(require(\\"path\\")).default;

var _universalImport = _interopRequireDefault(require(\\"babel-plugin-universal-import/universalImport\\")).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

_universalImport({
  id: \\"../../Foo\\",
  file: \\"currentFile.js\\",
  load: () => import(
  /* webpackChunkName: 'Foo' */
  \\"../../Foo\\"),
  path: () => _path.join(__dirname, \\"../../Foo\\"),
  resolve: () => require.resolveWeak(\\"../../Foo\\"),
  chunkName: () => \\"Foo\\"
});
"
`;

exports[`static import 1`] = `
"
import(\\"./Foo\\")

      ↓ ↓ ↓ ↓ ↓ ↓

var _path = _interopRequireDefault(require(\\"path\\")).default;

var _universalImport = _interopRequireDefault(require(\\"babel-plugin-universal-import/universalImport\\")).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

_universalImport({
  id: \\"./Foo\\",
  file: \\"currentFile.js\\",
  load: () => import(
  /* webpackChunkName: 'Foo' */
  \\"./Foo\\"),
  path: () => _path.join(__dirname, \\"./Foo\\"),
  resolve: () => require.resolveWeak(\\"./Foo\\"),
  chunkName: () => \\"Foo\\"
});
"
`;
